### PR TITLE
Alow complex domain credentials in basic auth. Fixes Issue #184

### PR DESCRIPTION
**Pull request by yakman2020**

### CODE REVIEW

File: Unix/http/httpclient.c [hunk 1]
```diff
@@ -2566,32 +2566,43 @@ MI_Result HttpClient_New_Connector2(
...
+                self->connector->username    = username;
+                self->connector->user_domain = NULL;
             }
-            else
+            else if ( authtype == AUTH_METHOD_NEGOTIATE ||
+                      authtype == AUTH_METHOD_NEGOTIATE_WITH_CREDS ||
+                      authtype == AUTH_METHOD_KERBEROS) 
             {
-                user_domain = memchr(username, '@', strlen(username));
+                user_domain = memchr(username, '\\', strlen(username));
```
Comment (paulcallen) [inline Unix/http/httpclient.c]: strchr is probably quicker as your solution always searches for the null terminator where strchr looks for the char until it hits the null terminator.

File: Unix/http/httpclient.c [hunk 2]
```diff
@@ -2566,32 +2566,43 @@ MI_Result HttpClient_New_Connector2(
...
-                    self->connector->user_domain = PAL_Strdup(user_domain+1);
+                    self->connector->username    = PAL_Strdup(user_domain+1);
+                    self->connector->user_domain = username;
                 }
                 else
                 {
-                    // No credential domain specified
-                    self->connector->username    = username;
-                    self->connector->user_domain = PAL_Strdup((char*)host);
+                    user_domain = memchr(username, '@', strlen(username));
```
Comment (paulcallen) [inline Unix/http/httpclient.c]: same comment about strchr

File: Unix/tests/http/test_httpclient.cpp [hunk 3]
```diff
@@ -141,6 +146,33 @@ void _StrandTestClientPost( _In_ Strand* self_, _In_ Message* msg )
     else
         s_charset = "";
 
+    if (request->headers->authorization)
```
Comment (paulcallen) [inline Unix/tests/http/test_httpclient.cpp]: aren't you going to want to crack open the basic auth header to validate it is encoded properly in there? That would be more complete.

**Comment by paulcallen** [comment]
can you add the issue number to the description so they are tied together.

File: Unix/tests/http/test_httpclient.cpp (hunk 3 above)
Comment (yakman2020) [inline Unix/tests/http/test_httpclient.cpp]: _StrandTestClientPost has a request which is coming from the server, after parsing of the http headers. I don't have access to the original auth header. But the password and username members of the HttpRequestMsg struct which I do get are busted and parsed exactly as the server is seeing them, so the work (which I actually built but couldn't use) wasn't necessary.

Of course that is only true about Basic.  For Nego we would have to figure something else out, but this isn't about Nego today

File: Unix/http/httpclient.c (hunk 2 above)
Comment (yakman2020) [inline Unix/http/httpclient.c]: done

File: Unix/http/httpclient.c (hunk 1 above)
Comment (yakman2020) [inline Unix/http/httpclient.c]: done

File: Unix/tests/http/test_httpclient.cpp (hunk 3 above)
Comment (paulcallen) [inline Unix/tests/http/test_httpclient.cpp]: ok, my bad

**Comment by paulcallen** [comment]
fixes issue #184

**Pull request merged**